src/build/cargo.rs

```diff
@@ -294,37 +293,25 @@ impl Executor for RlsExecutor {
             }
         }
 
-        let rustc_exe = env::var("RUSTC").unwrap_or(env::args().next().unwrap());
-        let mut cmd = Command::new(&rustc_exe);
+        // Cargo should handle `$RUSTC` wrapper override already
+        let mut cmd = cargo_cmd.clone();
         cmd.env(::RUSTC_SHIM_ENV_VAR_NAME, "1");
 
         // We only want to intercept rustc call targeting current crate to cache
         // args/envs generated by cargo so we can run only rustc later ourselves
         // Currently we don't cache nor modify build script args
-        let is_build_script = crate_name == "build_script_build";
+        let is_build_script = *target.kind() == TargetKind::CustomBuild;
         if !self.is_primary_crate(id) || is_build_script {
             let build_script_notice = if is_build_script {" (build script)"} else {""};
             trace!("rustc not intercepted - {}{}", id.name(), build_script_notice);
 
+            // Recreate the command, minus -Zsave-analysis.
             if ::CRATE_BLACKLIST.contains(&&*crate_name) {
-                // Recreate the command, minus -Zsave-analysis.
-                for a in cargo_args {
-                    if a != "-Zsave-analysis" {
-                        cmd.arg(a);
-                    }
-                }
-                cmd.envs(cargo_cmd.get_envs().iter().filter_map(|(k, v)| v.as_ref().map(|v| (k, v))));
-                if let Some(cwd) = cargo_cmd.get_cwd() {
-                    cmd.current_dir(cwd);
-                }
-                return match cmd.status() {
-                    Ok(ref e) if e.success() => Ok(()),
-                    _ => Err(CargoErrorKind::ProcessErrorKind(process_error(
-                                "process didn't exit successfully", None, None)).into()),
```

Comment:
We do - `cmd` is the clone of the `exec` argument, `cargo_cmd`, so it has it already